app/controllers/api/v1/deploys_controller.rb

```diff
@@ -0,0 +1,48 @@
+class Api::V1::DeploysController < Api::V1::ApplicationController
+  include CurrentProject
+
+  before_filter :authorize_project_deployer!
```

Comment:
Good point.
The `authorize_project_deployer!` could be refactored to a concern, but it uses different actions in both controllers so it might be more complicated than to explicitly call it here.

Regarding the reference logic, this is actually working differently in the main deploys controller.
In the UI you first show the changes of a branch and a warning if it was not found, but the deploys controller itself will not stop you from trying to deploy a non-existing branch (maybe this is desired...)

In the API, a request with a wrong `reference` will respond with 422.
It might be a good idea to add this error to the returned JSON (like `{ errors: { reference: ["does not exist"] } }`)

If you think we should change also the behavior of the main deploys controller to enforce the reference then I can move it into the `DeployService`.
The best way I know to prevent this kind of logic duplications is using a "form object" as the guys from [CodeClimate explain here](http://blog.codeclimate.com/blog/2012/10/17/7-ways-to-decompose-fat-activerecord-models/)
I wasn't sure if this is "inline" with the code style of the rest of the app.

WDYT?